lxd/device/device_common.go

```diff
@@ -92,3 +92,8 @@ func (d *deviceCommon) Update(oldDevices deviceConfig.Devices, isRunning bool) e
 func (d *deviceCommon) Remove() error {
 	return nil
 }
+
+// DevPath returns the device path on the host.
+func (d *deviceCommon) DevPath() string {
```

Comment:
As this is very QEMU specific and potentially error prone because the host path returned from Start() may be dynamic based on run-time configuration, how about storing the device paths returned from Start() in a json file in the instance's volume generated by the qemu driver, which can then be later be read back & edited, and used to regenerate the AppArmor profile?

Reply:
We can do that.

Reply:
There's one small issue if we removed `DevPath`. When a VM is running and LXD is restarted (VM is still running), there is no way for the newly started LXD server to know `devPaths` if the JSON file doesn't exist yet. If we then hotplug a disk, `devPaths` (and the JSON file) will only contain the disk path which is incorrect. A restart of the VM obviously fixes that.

Is this a problem we should consider?

Reply:
If its been deleted then we can expect problems but that seems unlikely. You can detect that and error before reloading aparmor though.

Reply:
This corner case would also happen after an update, so we shouldn't return an error in this case.

Reply:
What do you mean? You mean the first time lxd is updated to a version that supports this feature? If so then we have to error as the alternative is wiping out the apparmor Dev paths which isn't an option as it would break a running vm.

Reply:
Yes, that's what I meant. I would've liked a solution which works in the update case as well. Anyway, it now uses a `devpaths.json` which is required when updating devices.

Reply:
I see what you would prefer such an approach. And I'm not dead set against it at all. However my concern is that in its original form as a sort of "bolt-on" concept to the `device` type that it would easily get forgotten or broken when adding new devices or changing the behavior of existing ones, leading to strange bugs where editing one disk would then break another.

I think if we were going to use `DevPath()` as a concept then we would want to formalise the contract it gives to the instance drivers. By that I mean that we should then stop passing the device path via the RunConf returned from `Start()` and `Stop()` and instead update the instance drivers to call the device's `DevPath()` function after calling `Start()` or `Stop()`. This way should we ever find ourselves breaking that contract (i.e not being able to provide the `DevPath()` for a disk device outside of the `Start()` function) then we would immediately know about it as the instance wouldn't be able to start.

Also as `DevPath()` is rather disk specific, I suggest that rather than adding it to the main `Device` interface, we use interface-smuggling to create a new interface called `DeviceDisk` that provides `DevPath()` and that way one would be forced to implement it before the instance could use the device.

What do you think?

Reply:
I don't see why we would need to remove the device path from RunConf. That would just cause a lot of unnecessary code changes which are unrelated to this PR. The only reason I want a `DevPath` function is for when `devpaths.json` doesn't exist.

How about we just create a new interface `Disk` (as opposed to `DeviceDisk` as golint doesn't like that) for now which implements `DevPath`, and is implemented by disk and usb.

Reply:
> The only reason I want a `DevPath` function is for when `devpaths.json` doesn't exist.

We should not need the devpaths.json feature at all if we are implementing `DevPath` in the device.

However my concern is that `DevPath()` is only used populating the apparmor profile and isn't used for actually setting up the VM config. And because we are decoupling the returned `RunConf` (which is used in the QEMU config) from the `DevPath()` value (which is used for AppArmor profile) this leads to multiple potential problems down the line:

1. Firstly, there is the possibility that we may introduce a bug where the value of `DevPath()` and the dev path in `RunConf` differ for a device. We shouldn't have 2 sources of "truth" for a disk device's source device path.
2. By introducing `DevPath()` for AppArmor profile usage we are subtly changing the contract between instance and disk device drivers. Previously the contract was effectively that by calling device.Start() the device could return a start-time disk source device path in `RunConf` and the instance driver would use it. But now we are changing the contract to require the disk device to be able to provide the disk source device path at a non-start time (i.e `DevPath()` can be called at any time - and infact we rely on that because we are calling it for all disk devices at update time when another disk is hot-plugged).

Now this change of contract is fine (I'm assuming you've checked all scenarios of disk devices and all of them can provide a non-start time disk source device via `DevPath()`), but we should formalise it so that we don't accidentally introduce a change in the future that breaks that contract (perhaps a disk device mode that only knows the source of the disk device at start time due to some dynamic allocation of block devices) such that a non-start time call of `DevPath()` cannot be satified. In that case if we were not using `DevPath()` as the single source of truth for disk device setup then we likely wouldn't realise its a problem until we start getting strange issues due to apparmor profiles being changed for one disk device and removing the start-time added dev path for the dynamic disk device.

Also the way its was implemented by having all device types implement an empty `DevPath()` if not explicitly implemented meant that this too could be forgotten, whereas not have it implemented in the common driver and instead using interface smuggling will force the developer to remeber that `DevPath()` needs implementing.

Reply:
To help break the changes down, it would be good to have this `DevPaths` change as a separate PR, and this would also help to catch regressions.

Reply:
> Firstly, there is the possibility that we may introduce a bug where the value of DevPath() and the dev path in RunConf differ for a device. We shouldn't have 2 sources of "truth" for a disk device's source device path.

They will most likely differ, as in `DevPath` we only care about block devices. Everything else is ignored.

Reply:
I'm not sure I follow you I'm afraid. Are they not representing the same thing? (If not then `DevPath(`  should be named something else to avoid confusion).